File: platforms/android/contribute/src/main/java/pulse/contribute/android/util/Promise.kt
package pulse.contribute.android.util

class Promise(private val promiseRequest: String){

    companion object Static {
        //gets defined internally on App Start
        var promiseResolver: ((Any?, String) -> Unit)? = null
        var promiseUndefinedResolver: ((String) -> Unit)? = null
        var promiseObjectResolver: ((String, Any?, String) -> Unit)? = null
        var promiseException: ((String, String, String) -> Unit)? = null
    }

    fun resolve(returnValue: Any?) {
        promiseResolver?.invoke(returnValue, promiseRequest)
    }

    fun resolveUndefined() {
        promiseUndefinedResolver?.invoke(promiseRequest)
    }

    fun resolveTrue() {
        resolve(true)
    }

    fun resolveFalse() {
        resolve(false)
    }

    fun resolveObject(type: String, returnValue: Any?) {
        promiseObjectResolver?.invoke(type, returnValue, promiseRequest)
    }

    fun exception(name: String, message: String) {
        promiseException?.invoke("Native: $name", message, promiseRequest)
    }

}

File: platforms/dot-net/Promise.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PulseContribute.DotNet
{
    public class Promise
    {
        private string promiseId;

        public static Action<object, string>? _promiseResolver;
        public static Action<string>? _promiseUndefinedResolver;
        public static Action<string, object, string>? _promiseObjectResolver;
        public static Action<string, string, string>? _promiseException;

        public Promise(string promiseRequest)
        {
            promiseId = promiseRequest;
        }

        public static Promise Create(string promiseRequest)
        {
            return new Promise(promiseRequest);
        }

        public void Resolve(object returnValue)
        {
            _promiseResolver?.Invoke(returnValue, promiseId);
        }

        public void ResolveUndefined()
        {
            _promiseUndefinedResolver?.Invoke(promiseId);
        }

        public void ResolveTrue()
        {
            Resolve(true);
        }

        public void resolveFalse()
        {
            Resolve(false);
        }

        public void ResolveObject(string type, object returnValue)
        {
            _promiseObjectResolver?.Invoke(type, returnValue, promiseId);
        }

        public void Exception(string name, string message)
        {
            _promiseException?.Invoke("Native: " + name, message, promiseId);
        }

    }
}
